Tribler/community/tunnel/processes/subprocess.py

```diff
@@ -0,0 +1,192 @@
+
+        self.file_obj = file_obj
+        self.data_callback = data_callback
+        self.daemon = True
+        self.start()
+
+    def run(self):
+        """Keep consuming from the line until it is closed
+
+        :returns: None
+        """
+        line = ""
+        while not self.file_obj.closed:
+            try:
+                line += self.file_obj.readline()
+            except IOError:
+                break
+            if line.endswith('\n') and len(line) > 8:
+                line, data = unpack_complex(line)
+                if data is not None:
+                    reactor.callInThread(self.data_callback, data)
+
+
+class Subprocess(IProcess):
+
+    """The main entry-point handle: a subprocess object.
+        Overwritten by the subprocess for more advanced
+        functionality.
+    """
+
+    def __init__(self):
+        """Initialize a new Subprocess
+
+        :returns: None
+        """
+        super(Subprocess, self).__init__()
+
+        self.closed = Deferred()
+
+    def start(self):
+        """Start consuming from the input file descriptors
+
+        :returns: None
+        """
+        LineConsumer(FILE_CTRL_IN, self.on_ctrl)
+        LineConsumer(FILE_DATA_IN, self.on_data)
+        LineConsumer(FILE_EXIT_IN, self.on_exit)
+
+    def write_ctrl(self, s):
+        """Write a control message to the parent process
+
+        :param s: the message to send
+        :type s: str
+        :returns: None
+        """
+        Subprocess.write(FILE_CTRL_OUT, s, LOCK_CTRL)
+
+    def write_data(self, s):
+        """Write raw data to the parent process
+
+        :param s: the message to send
+        :type s: str
+        :returns: None
+        """
+        Subprocess.write(FILE_DATA_OUT, s, LOCK_DATA)
+
+    def write_exit(self, s):
+        """Write an exit message to the parent process
+
+        :param s: the message to send
+        :type s: str
+        :returns: None
+        """
+        Subprocess.write(FILE_EXIT_OUT, s, LOCK_EXIT)
+
+    @staticmethod
+    def close_all_streams():
+        """Close all registered file descriptors
+
+        :returns: None
+        """
+        # We use the fact that they are assigned
+        # to the range [3, 8].
+        for fno in range(3, 9, 1):
+            Subprocess.close(fno)
+
+    @staticmethod
+    def write(f, data, lock):
+        """Write to the parent process
+
+        :param f: the file to write to
+        :type f: file
+        :param data: the data to write
+        :type data: str
+        :param lock: the Lock to acquire
+        :type lock: threading.Lock
+        :returns: None
+        """
+        packed = pack_data(data)
+        def helper():
+            lock.acquire(True)
+            try:
+                f.write(packed)
+                f.flush()
+            except IOError:
+                pass
+            finally:
+                lock.release()
+        threading.Thread(target=helper).start()
```

Comment:
Why are you creating a thread every time you want to write something to the parent process? What about the overhead due to context switches?